### PR TITLE
Loop dashboard payment sheet changes

### PR DESCRIPTION
**Pull request by jediknightpython**
excel integration for generation payment sheets: single file downloaded containing multiple sheets
single download button for every sheet

### CODE REVIEW

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 1]
```diff
@@ -540,6 +540,9 @@ function change_graph(parameter) {
     }
 }
 
+
```
Comment (jediknightpython) [inline dg/media/app_dashboards/js/loop_dashboard.js]: done

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 2]
```diff
@@ -2821,36 +2855,72 @@ function aggregator_payment_sheet(data_json, aggregator) {
...
                     break;
                 }
             }
         }
     }
 
+
+    
+
+    var formatVal = function (yourNumber) {
```
Comment (jediknightpython) [inline dg/media/app_dashboards/js/loop_dashboard.js]: function removed from footercallback and defined globally once

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 3]
```diff
@@ -2821,36 +2855,72 @@ function aggregator_payment_sheet(data_json, aggregator) {
...
+    //Seperates the components of the number
+        var n= yourNumber.toString().split(".");
+    //Comma-fies the first part
+        if(n[0])
+            {
+                n[0] = n[0].replace(/\B(?=(\d{3})+(?!\d))/g, ",");
```
Comment (jediknightpython) [inline dg/media/app_dashboards/js/loop_dashboard.js]: regex for putting comma separators for numbers of form 1156 -> 1,156

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 4]
```diff
@@ -2859,27 +2929,151 @@ function aggregator_payment_sheet(data_json, aggregator) {
...
+                var a = n[1].toString();
+                n[1] = parseInt(a.charAt(0));
+            }
+
+    //Combines the two sections
+                return n.join(".");
+            }
+ 
+            // Total over all pages
+            total3 = api
```
Comment (jediknightpython) [inline dg/media/app_dashboards/js/loop_dashboard.js]: done

File: dg/templates/app_dashboards/loop_dashboard.html [hunk 5]
```diff
@@ -506,17 +510,60 @@ <h5 id="recent_sustainability_card"></h5>
 			</div>
 			<div class="row center">
 					<table id="table2" class="stripe">
+				<!-- 	<thead>
```
Comment (lokeshgarg92) [inline dg/templates/app_dashboards/loop_dashboard.html]: remove commented code

File: loop/payment_template.py [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+from models import LoopUser, CombinedTransaction, Village, Crop, Mandi, Farmer, DayTransportation, Gaddidar, Transporter, Language, CropLanguage, GaddidarCommission, GaddidarShareOutliers
+
+from loop_data_log import get_latest_timestamp
+from loop.payment_template import *
+# Create your views here.
+HELPLINE_NUMBER = "09891256494"
+
+TOTAL_NUMBER_OF_PRINTABLE_COLUMNS = 10
+NUMBER_OF_SHEETS = 3
+NAME_OF_SHEETS = ['Sheet1, Sheet2', 'Sheet3', 'Sheet4']
```
Comment (lokeshgarg92) [inline loop/payment_template.py]: use proper name of sheets instead sheet1,2,3,4

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 7]
```diff
@@ -2821,36 +2869,74 @@ function aggregator_payment_sheet(data_json, aggregator) {
...
                     break;
                 }
             }
         }
     }
 
+
+    
+
+    var formatVal = function (yourNumber) {
```
Comment (lokeshgarg92) [inline dg/media/app_dashboards/js/loop_dashboard.js]: this function is not required, javascript provide inbuilt functions to achieve the same. Please explore and do the needful

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 8]
```diff
@@ -2821,36 +2869,74 @@ function aggregator_payment_sheet(data_json, aggregator) {
...
+
+    var finalFormat = function (value){
+        if(value.indexOf('.') === -1)
+            return value + '.0';
+        else
+            return value;
+    }
+
+        // Remove the formatting to get integer data for summation
+    var intVal = function ( column_value ) {
```
Comment (lokeshgarg92) [inline dg/media/app_dashboards/js/loop_dashboard.js]: Please justify the use of intVal function, couldn't find any case where column_value is not 'number'

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 9]
```diff
@@ -2821,36 +2869,74 @@ function aggregator_payment_sheet(data_json, aggregator) {
...
+            title: "Aggregator Payment[AP] (in Rs) (0.25*Q)"
         }, {
-            title: "Transport Cost"
+            title: "Transport Cost[TC] (in Rs)"
         }, {
-            title: "Farmer Share"
+            title: "Farmers' Contribution[FC] (in Rs)"
         }, {
-            title: "Gaddidar Commission"
+            title: "Commision Agent Contribution[CAC] (in Rs)"
```
Comment (lokeshgarg92) [inline dg/media/app_dashboards/js/loop_dashboard.js]: please correct the spelling of commision to commission

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 10]
```diff
@@ -615,6 +618,50 @@ function set_filterlistener() {
...
+            return a.click();
+                }
+            };
+            xhttp.open("POST", "/loop/get_payment_sheet/", true);
+            xhttp.setRequestHeader("Content-Type", "application/json");
+            xhttp.responseType = 'blob';
+            var data_json = {
+                aggregator_data:{
+                    name: aggregator_sheet_name,
+                    data : data_set
```
Comment (jediknightpython) [inline dg/media/app_dashboards/js/loop_dashboard.js]: done

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 11]
```diff
@@ -2810,8 +2859,7 @@ function aggregator_payment_sheet(data_json, aggregator) {
     for (var i = 0; i < dates.length; i++) {
         for (var j = 0; j < mandis[i].length; j++) {
             var net_payment = (quantites[i][j] * AGGREGATOR_INCENTIVE_PERCENTAGE) + transport_cost[i][j] - farmer_share[i][j];
-
-            data_set.push([sno, dates[i], mandis[i][j], (quantites[i][j]).toString().concat(KG), farmers[i][j], (quantites[i][j] * AGGREGATOR_INCENTIVE_PERCENTAGE).toFixed(2), transport_cost[i][j], farmer_share[i][j], 0, net_payment]);
+            data_set.push([sno, dates[i], mandis[i][j], parseFloat(quantites[i][j].toFixed(2)), farmers[i][j], parseFloat((quantites[i][j] * AGGREGATOR_INCENTIVE_PERCENTAGE).toFixed(2)), transport_cost[i][j], farmer_share[i][j], 0, parseFloat(net_payment.toFixed(2))]);
```
Comment (jediknightpython) [inline dg/media/app_dashboards/js/loop_dashboard.js]: done

File: dg/media/app_dashboards/js/loop_dashboard.js (hunk 9 above)
Comment (jediknightpython) [inline dg/media/app_dashboards/js/loop_dashboard.js]: done

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 12]
```diff
@@ -2859,27 +2945,45 @@ function aggregator_payment_sheet(data_json, aggregator) {
...
-            title: "Gaddidar"
+            title: "Commision Agent"
         }, {
             title: "Market"
         }, {
-            title: "Quantity"
+            title: "Quantity[Q] (in Kg)"
         }, {
-            title: "Gaddidar Commission"
+            title: "Commision Agent Discount[CAD] (in Rs/Kg)"
```
Comment (jediknightpython) [inline dg/media/app_dashboards/js/loop_dashboard.js]: done

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 13]
```diff
@@ -2859,27 +2945,45 @@ function aggregator_payment_sheet(data_json, aggregator) {
...
         }, {
-            title: "Share"
+            title: "Commision Agent Contribution[CAC] (in Rs) (Q*CAD)"
```
Comment (jediknightpython) [inline dg/media/app_dashboards/js/loop_dashboard.js]: done

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 14]
```diff
@@ -2889,8 +2993,23 @@ function aggregator_payment_sheet(data_json, aggregator) {
                 "sExtends": "csv",
                 "sButtonText": "Download",
                 "bBomInc": true,
-                "sTitle": "Loop_" + getFormattedDate(aggregator) + "Gaddidar_Details"
+                "sTitle": "Loop_India_Bihar_Aggregator Payment_" + getFormattedDate(aggregator) + "Commision Agent Details"
```
Comment (jediknightpython) [inline dg/media/app_dashboards/js/loop_dashboard.js]: done

File: dg/media/app_dashboards/js/loop_dashboard.js [hunk 15]
```diff
@@ -2918,11 +3037,33 @@ function aggregator_payment_sheet(data_json, aggregator) {
...
+            $( api.column( 5 ).footer() ).html(
+                finalFormat(formatVal(total5))
+            );
+
         }
 
     });
+
+    aggregator_sheet_name = "Loop_India_Bihar_Aggregator Payment_" + getFormattedDate(aggregator) + "Payment Summary";
+    gaddidar_sheet_name = "Loop_India_Bihar_Aggregator Payment_" + getFormattedDate(aggregator) + "Commision Agent Details";
```
Comment (jediknightpython) [inline dg/media/app_dashboards/js/loop_dashboard.js]: done

File: dg/base_settings.py [hunk 16]
```diff
@@ -227,6 +227,23 @@
     'social.backends.google.GoogleOAuth2',
 )
 
+
+# LOGGING = {
```
Comment (jediknightpython) [inline dg/base_settings.py]: done

**Pull request merged**